### PR TITLE
Workaround Devcom-1559808 in ranges::sort

### PR DESCRIPTION
**Pull request by CaseyCarter**
If there's one thing that _must_ work in C++, it's sorting `vector` =)

For ease of review: DevCom-1559808.

This also moves `_Rewrap_subrange` to `<xutility>` because it's already being used there:
https://github.com/microsoft/STL/blob/ee5a216b516ef033d9153b324842369eee7653d1/stl/inc/xutility#L5674

### CODE REVIEW

File: tests/std/tests/P0896R4_ranges_alg_sort/test.cpp [hunk 1]
```diff
@@ -51,7 +52,31 @@ struct instantiator {
     }
 };
 
+constexpr void test_1559808() {
+    // Regression test for DevCom-1559808, a bad interaction between constexpr vector and the use of structured bindings
+    // in the implemenation of ranges::sort.
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_sort/test.cpp:57]: We implementors all live in an "implemenation". 🙄

File: stl/inc/algorithm [hunk 2]
```diff
@@ -179,34 +179,6 @@ namespace ranges {
         }
     };
 
-    template <class _Result, class _Wrapped, class _Unwrapped>
-    _NODISCARD constexpr _Result _Rewrap_subrange(_Wrapped& _Val, subrange<_Unwrapped>&& _UResult) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:183]: I wondered why this was moving to `<xutility>` until I saw that it's already being used there:

https://github.com/microsoft/STL/blob/ee5a216b516ef033d9153b324842369eee7653d1/stl/inc/xutility#L5674

No code change requested, but the PR description should mention this.

File: tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp [hunk 3]
```diff
@@ -71,9 +72,29 @@ constexpr void smoke_test() {
     }
 }
 
+constexpr void test_1559808() {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:75]: Although the comments mention DevCom, I recommend naming the test functions `test_devcom_1559808()`. We've gone through way too many bug databases, so plain numbers are ambiguous.

Occurs in other files.

File: tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp [hunk 4]
```diff
@@ -71,9 +72,29 @@ constexpr void smoke_test() {
     }
 }
 
+constexpr void test_1559808() {
+    // Regression test for DevCom-1559808, an interaction between vector and the
+    // use of structured bindings in the constexpr evaluator.
+
+    std::vector<int> haystack(33, 42), needle(8, 42); // No particular significance to any numbers in this function
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:79]: We conventionally avoid defining multiple variables on a single line.

Also occurs in `P0896R4_ranges_alg_search/test.cpp`.

File: tests/std/tests/P0896R4_ranges_alg_search/test.cpp [hunk 5]
```diff
@@ -168,8 +169,27 @@ constexpr bool run_tests() {
     return true;
 }
 
+constexpr void test_1559808() {
+    // Regression test for DevCom-1559808, an interaction between vector and the
+    // use of structured bindings in the constexpr evaluator.
+
+    std::vector<int> haystack(33, 42), needle(8, 42); // No particular significance to any numbers in this function
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_search/test.cpp:176]: This test is `using namespace std;`.

File: tests/std/tests/P0896R4_views_split/test.cpp [hunk 6]
```diff
@@ -320,7 +321,32 @@ constexpr bool instantiation_test() {
...
+    return true;
+}
+
 int main() {
     STATIC_ASSERT(instantiation_test());
     instantiation_test();
+
+#if defined(__clang__) || defined(__EDG__) // TRANSITION, DevDiv-1516290
+    STATIC_ASSERT(test_1559808());
+#endif // TRANSITION, DevDiv-1516290
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_views_split/test.cpp:350]: It appears that this intends to cite DevCom-1516290 "C++20 constexpr dynamic allocation doesn't properly understand std::construct_at()", not the much older DevDiv internal database.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
It looked like this was ready for review, so I went ahead and pushed minor changes. Please meow if you were still working on this! :cat:

File: tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp [hunk 7]
```diff
@@ -71,9 +72,30 @@ constexpr void smoke_test() {
     }
 }
 
+constexpr void test_devcom_1559808() {
+    // Regression test for DevCom-1559808, an interaction between vector and the
+    // use of structured bindings in the constexpr evaluator.
+
+    std::vector<int> haystack(33, 42); // No particular significance to any numbers in this function
```
Comment (miscco) [inline tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:79]: @STL I believe you missed those when pushing the review comments

Comment (cpplearner) [reply tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:79]: @miscco Do you mean it should not use `std::`? (The test is not `using namespace std;` though.)

Comment (miscco) [reply tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:79]: Yeah, there is no reason this one test should be special in that regard. We should rather add `using namespace std;` and remove the `std::` everywhere

Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_find_end/test.cpp:79]: According to:
```
casey@Concept:/mnt/c/STL/tests/std/tests$ grep -EL 'using[[:space:]]+namespace[[:space:]]+std' P0896R4*/test.cpp
P0896R4_P1614R2_comparisons/test.cpp
P0896R4_ranges_alg_adjacent_find/test.cpp
P0896R4_ranges_alg_equal/test.cpp
P0896R4_ranges_alg_fill/test.cpp
P0896R4_ranges_alg_fill_n/test.cpp
P0896R4_ranges_alg_find_end/test.cpp
P0896R4_ranges_alg_mismatch/test.cpp
P0896R4_ranges_iterator_machinery/test.cpp
P0896R4_ranges_to_address/test.cpp
```
there are nine P0896R4 tests that don't `using namespace std`. I will convert all of them in a different PR.

**Comment by CaseyCarter** [comment]
Reviewer note: I just inadvertently pushed a commit into this branch, and then force-pushed it back to e88474f. Hopefully this didn't reset incremental review.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed, or if more work is required.

**Comment by StephanTLavavej** [comment]
Thanks for fixing these bugs! :wrench: :lady_beetle: :heart_eyes_cat:

**Comment by mnatsuhara** [comment]
Did a post-commit review and all looks good!  (Just for the record 📝)

**Pull request merged**